test/unit/offset.js

```diff
@@ -568,4 +568,117 @@ QUnit.test( "iframe scrollTop/Left (see gh-1945)", function( assert ) {
 	}
 } );
 
+testIframe( "coordinates relative to document", "offset/rel-doc.html", function( assert, $, iframe, doc ) {
+	assert.expect( 768 );
+
+	var docElem = doc.documentElement, // <html> element
+		bodyElem = doc.body,
+		marker1 = doc.getElementById( "marker1" ),
+		marker2 = doc.getElementById( "marker2" ),
+		$marker1 = $( marker1 ),
+		$marker2 = $( marker2 ),
+
+		// Pixels
+		DOC_MARGIN = 1,
+		DOC_BORDER = 2,
+		DOC_PADDING = 4,
+		BODY_MARGIN = 8,
+		BODY_BORDER = 16,
+		BODY_PADDING = 32,
+		MARKER_TOP_LEFT = 64,
+		// Bit as Flag
+		FLAG_DOC_MARGIN = 1,
+		FLAG_DOC_BORDER = 2,
+		FLAG_DOC_PADDING = 4,
+		FLAG_BODY_MARGIN = 8,
+		FLAG_BODY_BORDER = 16,
+		FLAG_BODY_PADDING = 32;
+
+	// Test it in each situation that is made by combined properties.
+	function situations( affect, markerTopLeft, exBodyPadding, messageHead ) {
+		var flags, offset1, offset2, sumLen, labels, labelsInMessage;
+
+		function switchProp( elem, enable, styleProp, pixels, label ) {
+			if ( enable ) {
+				elem.style[ styleProp ] = pixels + "px";
+				if ( !exBodyPadding || elem !== bodyElem || styleProp !== "padding" ) {
+					sumLen += pixels;
+				}
+				labels.push( label );
+			} else {
+				elem.style[ styleProp ] = "0";
+			}
+		}
+
+		for ( flags = 0;
+				flags <= ( FLAG_DOC_MARGIN | FLAG_DOC_BORDER | FLAG_DOC_PADDING | FLAG_BODY_MARGIN | FLAG_BODY_BORDER | FLAG_BODY_PADDING );
+				flags++ ) {
```

Comment:
This is a very good start, but is a bit hard to read and hard to reason about. I now think every interesting case here can operate in a page with nonzero, non-pixel (to catch inappropriate `parseFloat`s) values for all of margin, border, and padding on both `<html>` and `<body>`, with the only variation being `position` values for those two elements. You can achieve that by means of a single iframe fixture (setting `position` styles before making any other calls), but I would like them in distinct `testIframe` invocations (see [effects.js](https://github.com/jquery/jquery/blob/master/test/unit/effects.js#L193) for an example of such process-constructed `QUnit.test` input). The cases, and my observations of them, are as follows:

`<html>` position | `<body>` position | `.offset` origin
------------ | ------------- | --------
static | static/relative/absolute/fixed | `document` (i.e., outer edge of `<html>` margin)
relative/absolute/fixed | static/relative/absolute/fixed | inside `<html>` border

That's a total of 16 cases, each case with its own `testIframe`. I'm aware that `<body>` position doesn't actually seem to matter, but it's probably worth covering anyway.